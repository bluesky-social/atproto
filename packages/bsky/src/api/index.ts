import { Server } from '../lexicon'
import AppContext from '../context'
import getTimeline from './app/bsky/feed/getTimeline'
import getAuthorFeed from './app/bsky/feed/getAuthorFeed'
import getLikes from './app/bsky/feed/getLikes'
import getPostThread from './app/bsky/feed/getPostThread'
import getProfile from './app/bsky/actor/getProfile'
import getProfiles from './app/bsky/actor/getProfiles'
import getRepostedBy from './app/bsky/feed/getRepostedBy'
import getFollowers from './app/bsky/graph/getFollowers'
import getFollows from './app/bsky/graph/getFollows'
import searchActors from './app/bsky/actor/searchActors'
import searchActorsTypeahead from './app/bsky/actor/searchActorsTypeahead'
import getSuggestions from './app/bsky/actor/getSuggestions'
import unspecced from './app/bsky/unspecced'

export * as health from './health'

export * as blobResolver from './blob-resolver'

export default function (server: Server, ctx: AppContext) {
  getTimeline(server, ctx)
  getAuthorFeed(server, ctx)
  getLikes(server, ctx)
  getPostThread(server, ctx)
  getProfile(server, ctx)
  getProfiles(server, ctx)
  getRepostedBy(server, ctx)
  getFollowers(server, ctx)
  getFollows(server, ctx)
  searchActors(server, ctx)
  searchActorsTypeahead(server, ctx)
  getSuggestions(server, ctx)
  unspecced(server, ctx)
  return server
}
