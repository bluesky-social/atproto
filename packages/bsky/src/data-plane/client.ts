import assert from 'node:assert'
import { randomInt } from 'node:crypto'
import * as ui8 from 'uint8arrays'
import {
  Code,
  ConnectError,
  PromiseClient,
  createPromiseClient,
  makeAnyClient,
} from '@connectrpc/connect'
import { createGrpcTransport } from '@connectrpc/connect-node'
import { getDidKeyFromMultibase } from '@atproto/identity'
import { Service } from '../proto/bsky_connect'

export type DataPlaneClient = PromiseClient<typeof Service>
type BaseClient = { lib: DataPlaneClient; url: URL }
type HttpVersion = '1.1' | '2'
const MAX_RETRIES = 3

export const createDataPlaneClient = (
  baseUrls: string[],
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
) => {
  const clients = baseUrls.map((baseUrl) => createBaseClient(baseUrl, opts))
  assert(clients.length > 0, 'no clients available')
  return makeAnyClient(Service, (method) => {
    return async (...args) => {
      let tries = 0
      let error: unknown
      let remainingClients = clients
      while (tries < MAX_RETRIES) {
        const client = randomElement(remainingClients)
        assert(client, 'no clients available')
        try {
          return await client.lib[method.localName](...args)
        } catch (err) {
          if (
            err instanceof ConnectError &&
            (err.code === Code.Unavailable || err.code === Code.Aborted)
          ) {
            tries++
            error = err
            remainingClients = getRemainingClients(remainingClients, client)
          } else {
            throw err
          }
        }
      }
      assert(error)
      throw error
    }
  }) as DataPlaneClient
}

export { Code }

export const isDataplaneError = (
  err: unknown,
  code?: Code,
): err is ConnectError => {
  if (err instanceof ConnectError) {
    return !code || err.code === code
  }
  return false
}

const createBaseClient = (
  baseUrl: string,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
): BaseClient => {
  const { httpVersion = '2', rejectUnauthorized = true } = opts
  const transport = createGrpcTransport({
    baseUrl,
    httpVersion,
    acceptCompression: [],
    nodeOptions: { rejectUnauthorized },
  })
  return {
    lib: createPromiseClient(Service, transport),
    url: new URL(baseUrl),
  }
}

const getRemainingClients = (clients: BaseClient[], lastClient: BaseClient) => {
  if (clients.length < 2) return clients // no clients to choose from
  if (lastClient.url.port) {
    // if the last client had a port, we attempt to exclude its whole host.
    const maybeRemaining = clients.filter(
      (c) => c.url.hostname !== lastClient.url.hostname,
    )
    if (maybeRemaining.length) {
      return maybeRemaining
    }
  }
  return clients.filter((c) => c !== lastClient)
}

const randomElement = <T>(arr: T[]): T | undefined => {
  if (arr.length === 0) return
  return arr[randomInt(arr.length)]
}

export const unpackIdentityServices = (servicesBytes: Uint8Array) => {
  const servicesStr = ui8.toString(servicesBytes, 'utf8')
  if (!servicesStr) return {}
  return JSON.parse(servicesStr) as UnpackedServices
}

export const unpackIdentityKeys = (keysBytes: Uint8Array) => {
  const keysStr = ui8.toString(keysBytes, 'utf8')
  if (!keysStr) return {}
  return JSON.parse(keysStr) as UnpackedKeys
}

export const getServiceEndpoint = (
  services: UnpackedServices,
  opts: { id: string; type: string },
) => {
  const endpoint =
    services[opts.id] &&
    services[opts.id].Type === opts.type &&
    validateUrl(services[opts.id].URL)
  return endpoint || undefined
}

export const getKeyAsDidKey = (keys: UnpackedKeys, opts: { id: string }) => {
  const key =
    keys[opts.id] &&
    getDidKeyFromMultibase({
      type: keys[opts.id].Type,
      publicKeyMultibase: keys[opts.id].PublicKeyMultibase,
    })
  return key || undefined
}

type UnpackedServices = Record<string, { Type: string; URL: string }>

type UnpackedKeys = Record<string, { Type: string; PublicKeyMultibase: string }>

const validateUrl = (urlStr: string): string | undefined => {
  let url
  try {
    url = new URL(urlStr)
  } catch {
    return undefined
  }
  if (!['http:', 'https:'].includes(url.protocol)) {
    return undefined
  } else if (!url.hostname) {
    return undefined
  } else {
    return urlStr
  }
}
