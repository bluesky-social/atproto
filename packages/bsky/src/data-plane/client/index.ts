import assert from 'node:assert'
import { randomInt } from 'node:crypto'
import {
  Code,
  ConnectError,
  PromiseClient,
  createPromiseClient,
  makeAnyClient,
} from '@connectrpc/connect'
import { createGrpcTransport } from '@connectrpc/connect-node'
import { Service } from '../../proto/bsky_connect'
import { HostList } from './hosts'

export * from './hosts'
export * from './util'

export type DataPlaneClient = PromiseClient<typeof Service>
type HttpVersion = '1.1' | '2'
const MAX_RETRIES = 3

export const createDataPlaneClient = (
  hostList: HostList,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
) => {
  const clients = new DataPlaneClients(hostList, opts)
  return makeAnyClient(Service, (method) => {
    return async (...args) => {
      let tries = 0
      let error: unknown
      let remainingClients = clients.get()
      while (tries < MAX_RETRIES) {
        const client = randomElement(remainingClients)
        assert(client, 'no clients available')
        try {
          return await client[method.localName](...args)
        } catch (err) {
          if (
            err instanceof ConnectError &&
            (err.code === Code.Unavailable || err.code === Code.Aborted)
          ) {
            tries++
            error = err
            remainingClients = getRemainingClients(remainingClients, client)
          } else {
            throw err
          }
        }
      }
      assert(error)
      throw error
    }
  }) as DataPlaneClient
}

export { Code }

/**
 * Uses a reactive HostList in order to maintain a pool of DataPlaneClients.
 * Each DataPlaneClient is cached per host so that it maintains connections
 * and other internal state when the underlying HostList is updated.
 */
class DataPlaneClients {
  private clients: DataPlaneClient[] = []
  private clientsByHost = new Map<string, DataPlaneClient>()

  constructor(
    private hostList: HostList,
    private clientOpts: {
      httpVersion?: HttpVersion
      rejectUnauthorized?: boolean
    },
  ) {
    this.refresh()
    this.hostList.onUpdate(() => this.refresh())
  }

  get(): readonly DataPlaneClient[] {
    return this.clients
  }

  private refresh() {
    this.clients = []
    for (const host of this.hostList.get()) {
      let client = this.clientsByHost.get(host)
      if (!client) {
        client = this.createClient(host)
        this.clientsByHost.set(host, client)
      }
      this.clients.push(client)
    }
  }

  private createClient(host: string) {
    return createBaseClient(host, this.clientOpts)
  }
}

const createBaseClient = (
  baseUrl: string,
  opts: { httpVersion?: HttpVersion; rejectUnauthorized?: boolean },
): DataPlaneClient => {
  const { httpVersion = '2', rejectUnauthorized = true } = opts
  const transport = createGrpcTransport({
    baseUrl,
    httpVersion,
    acceptCompression: [],
    nodeOptions: { rejectUnauthorized },
  })
  return createPromiseClient(Service, transport)
}

const getRemainingClients = (
  clients: readonly DataPlaneClient[],
  lastClient: DataPlaneClient,
) => {
  if (clients.length < 2) return clients // no clients to choose from
  return clients.filter((c) => c !== lastClient)
}

const randomElement = <T>(arr: readonly T[]): T | undefined => {
  if (arr.length === 0) return
  return arr[randomInt(arr.length)]
}
