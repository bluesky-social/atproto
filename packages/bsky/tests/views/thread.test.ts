import AtpAgent, { AppBskyFeedGetPostThread } from '@atproto/api'
import { Database } from '../../src'
import {
  runTestServer,
  forSnapshot,
  CloseFn,
  processAll,
  TestServerInfo,
  stripViewerFromThread,
} from '../_util'
import { RecordRef, SeedClient } from '../seeds/client'
import basicSeed from '../seeds/basic'
import threadSeed, { walk, item, Item } from '../seeds/thread'

describe('pds thread views', () => {
  let server: TestServerInfo
  let agent: AtpAgent
  let db: Database
  let close: CloseFn
  let sc: SeedClient

  // account dids, for convenience
  let alice: string
  let bob: string

  beforeAll(async () => {
    server = await runTestServer({
      dbPostgresSchema: 'views_thread',
    })
    db = server.ctx.db
    close = server.close
    agent = new AtpAgent({ service: server.url })
    const pdsAgent = new AtpAgent({ service: server.pdsUrl })
    sc = new SeedClient(pdsAgent)
    await basicSeed(sc)
    alice = sc.dids.alice
    bob = sc.dids.bob
  })

  beforeAll(async () => {
    // Add a repost of a reply so that we can confirm myState in the thread
    await sc.repost(bob, sc.replies[alice][0].ref)
    await processAll(server)
  })

  afterAll(async () => {
    await close()
  })

  // @TODO(bsky) blocks post by actor takedown via labels.
  // @TODO(bsky) blocks post by record takedown via labels.
  // @TODO(bsky) blocks replies by actor takedown via labels.
  // @TODO(bsky) blocks ancestors by record takedown via labels.
  // @TODO(bsky) blocks ancestors by actor takedown via labels.

  it('fetches deep post thread', async () => {
    const thread = await agent.api.app.bsky.feed.getPostThread(
      { uri: sc.posts[alice][1].ref.uriStr },
      { headers: sc.getHeaders(bob, true) },
    )

    expect(forSnapshot(thread.data.thread)).toMatchSnapshot()
  })

  it('fetches shallow post thread', async () => {
    const thread = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[alice][1].ref.uriStr },
      { headers: sc.getHeaders(bob, true) },
    )

    expect(forSnapshot(thread.data.thread)).toMatchSnapshot()
  })

  it('fetches ancestors', async () => {
    const thread = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.replies[alice][0].ref.uriStr },
      { headers: sc.getHeaders(bob, true) },
    )

    expect(forSnapshot(thread.data.thread)).toMatchSnapshot()
  })

  it('fails for an unknown post', async () => {
    const promise = agent.api.app.bsky.feed.getPostThread(
      { uri: 'at://did:example:fake/does.not.exist/self' },
      { headers: sc.getHeaders(bob, true) },
    )

    await expect(promise).rejects.toThrow(
      AppBskyFeedGetPostThread.NotFoundError,
    )
  })

  it('fetches post thread unauthed', async () => {
    const { data: authed } = await agent.api.app.bsky.feed.getPostThread(
      { uri: sc.posts[alice][1].ref.uriStr },
      { headers: sc.getHeaders(bob, true) },
    )
    const { data: unauthed } = await agent.api.app.bsky.feed.getPostThread({
      uri: sc.posts[alice][1].ref.uriStr,
    })
    expect(unauthed.thread).toEqual(stripViewerFromThread(authed.thread))
  })

  it('handles deleted posts correctly', async () => {
    const alice = sc.dids.alice
    const bob = sc.dids.bob

    const indexes = {
      aliceRoot: -1,
      bobReply: -1,
      aliceReplyReply: -1,
    }

    await sc.post(alice, 'Deletion thread')
    indexes.aliceRoot = sc.posts[alice].length - 1

    await sc.reply(
      bob,
      sc.posts[alice][indexes.aliceRoot].ref,
      sc.posts[alice][indexes.aliceRoot].ref,
      'Reply',
    )
    indexes.bobReply = sc.replies[bob].length - 1
    await sc.reply(
      alice,
      sc.posts[alice][indexes.aliceRoot].ref,
      sc.replies[bob][indexes.bobReply].ref,
      'Reply reply',
    )
    indexes.aliceReplyReply = sc.replies[alice].length - 1
    await processAll(server)

    const thread1 = await agent.api.app.bsky.feed.getPostThread(
      { uri: sc.posts[alice][indexes.aliceRoot].ref.uriStr },
      { headers: sc.getHeaders(bob, true) },
    )
    expect(forSnapshot(thread1.data.thread)).toMatchSnapshot()

    await sc.deletePost(bob, sc.replies[bob][indexes.bobReply].ref.uri)
    await processAll(server)

    const thread2 = await agent.api.app.bsky.feed.getPostThread(
      { uri: sc.posts[alice][indexes.aliceRoot].ref.uriStr },
      { headers: sc.getHeaders(bob, true) },
    )
    expect(forSnapshot(thread2.data.thread)).toMatchSnapshot()

    const thread3 = await agent.api.app.bsky.feed.getPostThread(
      { uri: sc.replies[alice][indexes.aliceReplyReply].ref.uriStr },
      { headers: sc.getHeaders(bob, true) },
    )
    expect(forSnapshot(thread3.data.thread)).toMatchSnapshot()
  })

  it('builds post hierarchy index.', async () => {
    const threads: Item[] = [
      item(1, [item(2, [item(3), item(4)])]),
      item(5, [item(6), item(7, [item(9, [item(11)]), item(10)]), item(8)]),
      item(12),
    ]

    await threadSeed(sc, sc.dids.alice, threads)
    await processAll(server)

    let closureSize = 0
    const itemByUri: Record<string, Item> = {}

    const postsAndReplies = ([] as { text: string; ref: RecordRef }[])
      .concat(Object.values(sc.posts[sc.dids.alice]))
      .concat(Object.values(sc.replies[sc.dids.alice]))
      .filter((p) => {
        const id = parseInt(p.text, 10)
        return 0 < id && id <= 12
      })

    await walk(threads, async (item, depth) => {
      const post = postsAndReplies.find((p) => p.text === String(item.id))
      if (!post) throw new Error('Post not found')
      itemByUri[post.ref.uriStr] = item
      closureSize += depth + 1
    })

    const hierarchy = await db.db
      .selectFrom('post_hierarchy')
      .where(
        'uri',
        'in',
        postsAndReplies.map((p) => p.ref.uriStr),
      )
      .orWhere(
        'ancestorUri',
        'in',
        postsAndReplies.map((p) => p.ref.uriStr),
      )
      .selectAll()
      .execute()

    expect(hierarchy.length).toEqual(closureSize)

    for (const relation of hierarchy) {
      const item = itemByUri[relation.uri]
      const ancestor = itemByUri[relation.ancestorUri]
      let depth = -1
      await walk([ancestor], async (candidate, candidateDepth) => {
        if (candidate === item) {
          depth = candidateDepth
        }
      })
      expect(depth).toEqual(relation.depth)
    }
  })
})
