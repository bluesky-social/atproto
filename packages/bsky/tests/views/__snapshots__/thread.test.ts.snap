// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pds thread views fetches ancestors 1`] = `
Object {
  "$type": "app.bsky.feed.defs#threadViewPost",
  "parent": Object {
    "$type": "app.bsky.feed.defs#threadViewPost",
    "parent": Object {
      "$type": "app.bsky.feed.defs#threadViewPost",
      "post": Object {
        "author": Object {
          "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
          "did": "user(0)",
          "displayName": "ali",
          "handle": "alice.test",
          "viewer": Object {
            "followedBy": "record(2)",
            "following": "record(1)",
          },
        },
        "cid": "cids(2)",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "likeCount": 3,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "text": "again",
        },
        "replyCount": 2,
        "repostCount": 1,
        "uri": "record(3)",
        "viewer": Object {
          "like": "record(6)",
        },
      },
      "replies": Array [],
    },
    "post": Object {
      "author": Object {
        "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(3)/cids(1)@jpeg",
        "did": "user(2)",
        "displayName": "bobby",
        "handle": "bob.test",
        "viewer": Object {},
      },
      "cid": "cids(3)",
      "embed": Object {
        "$type": "app.bsky.embed.images#view",
        "images": Array [
          Object {
            "alt": "tests/image/fixtures/key-landscape-small.jpg",
            "fullsize": "https://bsky.public.url/image/sig()/rs:fit:2000:2000:1:0/plain/user(3)/cids(4)@jpeg",
            "thumb": "https://bsky.public.url/image/sig()/rs:fit:1000:1000:1:0/plain/user(3)/cids(4)@jpeg",
          },
        ],
      },
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "likeCount": 0,
      "record": Object {
        "$type": "app.bsky.feed.post",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "embed": Object {
          "$type": "app.bsky.embed.images",
          "images": Array [
            Object {
              "alt": "tests/image/fixtures/key-landscape-small.jpg",
              "image": Object {
                "$type": "blob",
                "mimeType": "image/jpeg",
                "ref": Object {
                  "$link": "cids(4)",
                },
                "size": 4114,
              },
            },
          ],
        },
        "reply": Object {
          "parent": Object {
            "cid": "cids(2)",
            "uri": "record(3)",
          },
          "root": Object {
            "cid": "cids(2)",
            "uri": "record(3)",
          },
        },
        "text": "hear that",
      },
      "replyCount": 1,
      "repostCount": 0,
      "uri": "record(4)",
      "viewer": Object {},
    },
    "replies": Array [],
  },
  "post": Object {
    "author": Object {
      "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "viewer": Object {
        "followedBy": "record(2)",
        "following": "record(1)",
      },
    },
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "likeCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "reply": Object {
        "parent": Object {
          "cid": "cids(3)",
          "uri": "record(4)",
        },
        "root": Object {
          "cid": "cids(2)",
          "uri": "record(3)",
        },
      },
      "text": "thanks bob",
    },
    "replyCount": 0,
    "repostCount": 1,
    "uri": "record(0)",
    "viewer": Object {
      "repost": "record(5)",
    },
  },
  "replies": Array [],
}
`;

exports[`pds thread views fetches deep post thread 1`] = `
Object {
  "$type": "app.bsky.feed.defs#threadViewPost",
  "post": Object {
    "author": Object {
      "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "viewer": Object {
        "followedBy": "record(2)",
        "following": "record(1)",
      },
    },
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "likeCount": 3,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "again",
    },
    "replyCount": 2,
    "repostCount": 1,
    "uri": "record(0)",
    "viewer": Object {
      "like": "record(3)",
    },
  },
  "replies": Array [
    Object {
      "$type": "app.bsky.feed.defs#threadViewPost",
      "post": Object {
        "author": Object {
          "did": "user(2)",
          "handle": "carol.test",
          "viewer": Object {
            "following": "record(5)",
          },
        },
        "cid": "cids(2)",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "reply": Object {
            "parent": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
            "root": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
          },
          "text": "of course",
        },
        "replyCount": 0,
        "repostCount": 0,
        "uri": "record(4)",
        "viewer": Object {},
      },
      "replies": Array [],
    },
    Object {
      "$type": "app.bsky.feed.defs#threadViewPost",
      "post": Object {
        "author": Object {
          "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(4)/cids(1)@jpeg",
          "did": "user(3)",
          "displayName": "bobby",
          "handle": "bob.test",
          "viewer": Object {},
        },
        "cid": "cids(3)",
        "embed": Object {
          "$type": "app.bsky.embed.images#view",
          "images": Array [
            Object {
              "alt": "tests/image/fixtures/key-landscape-small.jpg",
              "fullsize": "https://bsky.public.url/image/sig()/rs:fit:2000:2000:1:0/plain/user(4)/cids(4)@jpeg",
              "thumb": "https://bsky.public.url/image/sig()/rs:fit:1000:1000:1:0/plain/user(4)/cids(4)@jpeg",
            },
          ],
        },
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "embed": Object {
            "$type": "app.bsky.embed.images",
            "images": Array [
              Object {
                "alt": "tests/image/fixtures/key-landscape-small.jpg",
                "image": Object {
                  "$type": "blob",
                  "mimeType": "image/jpeg",
                  "ref": Object {
                    "$link": "cids(4)",
                  },
                  "size": 4114,
                },
              },
            ],
          },
          "reply": Object {
            "parent": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
            "root": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
          },
          "text": "hear that",
        },
        "replyCount": 1,
        "repostCount": 0,
        "uri": "record(6)",
        "viewer": Object {},
      },
      "replies": Array [
        Object {
          "$type": "app.bsky.feed.defs#threadViewPost",
          "post": Object {
            "author": Object {
              "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
              "did": "user(0)",
              "displayName": "ali",
              "handle": "alice.test",
              "viewer": Object {
                "followedBy": "record(2)",
                "following": "record(1)",
              },
            },
            "cid": "cids(5)",
            "indexedAt": "1970-01-01T00:00:00.000Z",
            "likeCount": 0,
            "record": Object {
              "$type": "app.bsky.feed.post",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "reply": Object {
                "parent": Object {
                  "cid": "cids(3)",
                  "uri": "record(6)",
                },
                "root": Object {
                  "cid": "cids(0)",
                  "uri": "record(0)",
                },
              },
              "text": "thanks bob",
            },
            "replyCount": 0,
            "repostCount": 1,
            "uri": "record(7)",
            "viewer": Object {
              "repost": "record(8)",
            },
          },
          "replies": Array [],
        },
      ],
    },
  ],
}
`;

exports[`pds thread views fetches shallow post thread 1`] = `
Object {
  "$type": "app.bsky.feed.defs#threadViewPost",
  "post": Object {
    "author": Object {
      "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "viewer": Object {
        "followedBy": "record(2)",
        "following": "record(1)",
      },
    },
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "likeCount": 3,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "again",
    },
    "replyCount": 2,
    "repostCount": 1,
    "uri": "record(0)",
    "viewer": Object {
      "like": "record(3)",
    },
  },
  "replies": Array [
    Object {
      "$type": "app.bsky.feed.defs#threadViewPost",
      "post": Object {
        "author": Object {
          "did": "user(2)",
          "handle": "carol.test",
          "viewer": Object {
            "following": "record(5)",
          },
        },
        "cid": "cids(2)",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "reply": Object {
            "parent": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
            "root": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
          },
          "text": "of course",
        },
        "replyCount": 0,
        "repostCount": 0,
        "uri": "record(4)",
        "viewer": Object {},
      },
    },
    Object {
      "$type": "app.bsky.feed.defs#threadViewPost",
      "post": Object {
        "author": Object {
          "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(4)/cids(1)@jpeg",
          "did": "user(3)",
          "displayName": "bobby",
          "handle": "bob.test",
          "viewer": Object {},
        },
        "cid": "cids(3)",
        "embed": Object {
          "$type": "app.bsky.embed.images#view",
          "images": Array [
            Object {
              "alt": "tests/image/fixtures/key-landscape-small.jpg",
              "fullsize": "https://bsky.public.url/image/sig()/rs:fit:2000:2000:1:0/plain/user(4)/cids(4)@jpeg",
              "thumb": "https://bsky.public.url/image/sig()/rs:fit:1000:1000:1:0/plain/user(4)/cids(4)@jpeg",
            },
          ],
        },
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "embed": Object {
            "$type": "app.bsky.embed.images",
            "images": Array [
              Object {
                "alt": "tests/image/fixtures/key-landscape-small.jpg",
                "image": Object {
                  "$type": "blob",
                  "mimeType": "image/jpeg",
                  "ref": Object {
                    "$link": "cids(4)",
                  },
                  "size": 4114,
                },
              },
            ],
          },
          "reply": Object {
            "parent": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
            "root": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
          },
          "text": "hear that",
        },
        "replyCount": 1,
        "repostCount": 0,
        "uri": "record(6)",
        "viewer": Object {},
      },
    },
  ],
}
`;

exports[`pds thread views handles deleted posts correctly 1`] = `
Object {
  "$type": "app.bsky.feed.defs#threadViewPost",
  "post": Object {
    "author": Object {
      "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "viewer": Object {
        "followedBy": "record(2)",
        "following": "record(1)",
      },
    },
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "likeCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "Deletion thread",
    },
    "replyCount": 1,
    "repostCount": 0,
    "uri": "record(0)",
    "viewer": Object {},
  },
  "replies": Array [
    Object {
      "$type": "app.bsky.feed.defs#threadViewPost",
      "post": Object {
        "author": Object {
          "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(3)/cids(1)@jpeg",
          "did": "user(2)",
          "displayName": "bobby",
          "handle": "bob.test",
          "viewer": Object {},
        },
        "cid": "cids(2)",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "reply": Object {
            "parent": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
            "root": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
          },
          "text": "Reply",
        },
        "replyCount": 1,
        "repostCount": 0,
        "uri": "record(3)",
        "viewer": Object {},
      },
      "replies": Array [
        Object {
          "$type": "app.bsky.feed.defs#threadViewPost",
          "post": Object {
            "author": Object {
              "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
              "did": "user(0)",
              "displayName": "ali",
              "handle": "alice.test",
              "viewer": Object {
                "followedBy": "record(2)",
                "following": "record(1)",
              },
            },
            "cid": "cids(3)",
            "indexedAt": "1970-01-01T00:00:00.000Z",
            "likeCount": 0,
            "record": Object {
              "$type": "app.bsky.feed.post",
              "createdAt": "1970-01-01T00:00:00.000Z",
              "reply": Object {
                "parent": Object {
                  "cid": "cids(2)",
                  "uri": "record(3)",
                },
                "root": Object {
                  "cid": "cids(0)",
                  "uri": "record(0)",
                },
              },
              "text": "Reply reply",
            },
            "replyCount": 0,
            "repostCount": 0,
            "uri": "record(4)",
            "viewer": Object {},
          },
          "replies": Array [],
        },
      ],
    },
  ],
}
`;

exports[`pds thread views handles deleted posts correctly 2`] = `
Object {
  "$type": "app.bsky.feed.defs#threadViewPost",
  "post": Object {
    "author": Object {
      "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "viewer": Object {
        "followedBy": "record(2)",
        "following": "record(1)",
      },
    },
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "likeCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "text": "Deletion thread",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(0)",
    "viewer": Object {},
  },
  "replies": Array [],
}
`;

exports[`pds thread views handles deleted posts correctly 3`] = `
Object {
  "$type": "app.bsky.feed.defs#threadViewPost",
  "parent": Object {
    "$type": "app.bsky.feed.defs#notFoundPost",
    "notFound": true,
    "uri": "record(4)",
  },
  "post": Object {
    "author": Object {
      "avatar": "https://bsky.public.url/image/sig()/rs:fill:1000:1000:1:0/plain/user(1)/cids(1)@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "viewer": Object {
        "followedBy": "record(2)",
        "following": "record(1)",
      },
    },
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "likeCount": 0,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000Z",
      "reply": Object {
        "parent": Object {
          "cid": "cids(3)",
          "uri": "record(4)",
        },
        "root": Object {
          "cid": "cids(2)",
          "uri": "record(3)",
        },
      },
      "text": "Reply reply",
    },
    "replyCount": 0,
    "repostCount": 0,
    "uri": "record(0)",
    "viewer": Object {},
  },
  "replies": Array [],
}
`;
